src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java

```diff
@@ -0,0 +1,271 @@
+     * able to write to storage
+     */
+    @Test
+    public void testThatSettingJWTKeysInParallelWorksAsExpectedForNoSQL() throws Exception {
+        String[] args = {"../"};
+        TestingProcessManager.TestingProcess process = TestingProcessManager.start(args);
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STARTED));
+
+        for (int i = 0; i < 10; i++) {
+            String algorithm = "alg" + i;
+            String keyId = "key" + i;
+
+            JWTRecipeStorage storage = StorageLayer.getJWTRecipeStorage(process.getProcess());
+
+            if (storage.getType() != STORAGE_TYPE.NOSQL_1) {
+                return;
+            }
+
+            JWTRecipeNoSQLStorage_1 noSQLStorage_1 = (JWTRecipeNoSQLStorage_1) storage;
+
+            AtomicReference<String> endValueOfCon1 = new AtomicReference<>("c1");
+            AtomicReference<String> endValueOfCon2 = new AtomicReference<>("c2");
+            AtomicInteger numberOfIterations = new AtomicInteger(0);
+
+            Runnable runnable1 = new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        while (true) {
+                            numberOfIterations.getAndIncrement();
+                            List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage_1.getJWTSigningKeys_Transaction();
+
+                            JWTSigningKeyInfo matchingKeyInfo = null;
+
+                            for (int j = 0; j < keysFromStorage.size(); j ++) {
+                                if (keysFromStorage.get(j).algorithm.equals(algorithm)) {
+                                    matchingKeyInfo = keysFromStorage.get(j);
+                                    break;
+                                }
+                            }
+
+                            try {
+                                Thread.sleep(300);
+                            } catch (InterruptedException e) {
+                                // Ignore
+                            }
+
+                            try {
+                                if (matchingKeyInfo == null) {
+                                    boolean success = noSQLStorage_1.setJWTSigningKeyInfoIfNoKeyForAlgorithmExists_Transaction(new JWTSymmetricSigningKeyInfo(keyId + "-1", 1000, algorithm, "SomeKeyString"));
+
+                                    if (!success) {
+                                        fail();
+                                    }
+
+                                    endValueOfCon1.set(keyId + "-1");
+                                } else {
+                                    endValueOfCon1.set(matchingKeyInfo.keyId);
+                                }
+
+                                break;
+                            } catch (DuplicateKeyIdException e) {
+                                continue;
```

Comment:
Throw an exception here.

Reply:
Done, the test has been refactored. All exceptions that are not expected result in the test failing.